### PR TITLE
Sampler

### PR DESCRIPTION
**Pull request by fgrunewald**
Initial draft for a sampler class that takes fragments and makes random molecules according to bonding probabilities

### CODE REVIEW

File: cgsmiles/sample.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            for node, bondings in bondings.items():
+                for bonding in bondings:
+                    self.fragments_by_bonding[bonding].append((fragname, node))
+
+    def update_open_bonds(self):
+        self.current_open_bonds = defaultdict(list)
+        open_bonds = nx.get_node_attributes(self.molecule, 'bonding')
+        for node, bonding_types in open_bonds.items():
+            for bonding_types in bonding_types:
+                self.current_open_bonds[bonding_types].append(node)
```
Comment (pckroon) [inline cgsmiles/sample.py:83]: Is it worth abstracting this to a helper function?

File: cgsmiles/sample.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        merge_graphs(self.molecule, fragment)
+        self.update_open_bonds()
+
+        # next we add monomers one after the other
+        for _ in np.arange(0, self.target_weight):
+            # pick a bonding connector
+            source, target, fragname, bonding, compl_bonding = self.pick_bonding()
+            correspondence = merge_graphs(self.molecule, self.fragment_dict[fragname])
+            self.molecule.add_edge(source, correspondence[target], bonding=(bonding, compl_bonding))
+            print(source, bonding, self.molecule.nodes[source]['bonding'])
```
Comment (pckroon) [inline cgsmiles/sample.py:119]:
```suggestion
```

**Review by pckroon — CHANGES_REQUESTED** [review]
Some small comments, looks mostly fine to me.

File: cgsmiles/sample.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        fragment_probabilities: dict[str, float]
+            dictionary connecting probability
+            to fragment names
+        bonding_probabilities: dict[str, float]
+            probabilty that two bonding descriptors
+            will connect
+        all_atom: bool
+            if the fragments are all-atom
+            resolution
+        """
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/sample.py:52]: @pckroon we could roll the same way as with the MoleculeResolver, and have a constructor.

File: cgsmiles/sample.py (hunk 1 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:83]: I guess could be useful; why not

File: cgsmiles/sample.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    def pick_bonding(self):
+        """
+        Given a momomer to connect to the growing molecule see if there is
+        a matching bonding descriptor and what the probabilitiy of forming
+        this bond is.
+        """
+        probs = [self.bonding_probabilities[bond_type] for bond_type in self.current_open_bonds]
+        bonding = np.random.choice(list(self.current_open_bonds.keys()), p=probs)
+        source_node = np.random.choice(self.current_open_bonds[bonding])
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/sample.py:93]: random should do the trick

File: cgsmiles/sample.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+        merge_graphs(self.molecule, fragment)
+        self.update_open_bonds()
+
+        # next we add monomers one after the other
+        for _ in np.arange(0, self.target_weight):
+            # pick a bonding connector
+            source, target, fragname, bonding, compl_bonding = self.pick_bonding()
+            correspondence = merge_graphs(self.molecule, self.fragment_dict[fragname])
+            self.molecule.add_edge(source, correspondence[target], bonding=(bonding, compl_bonding))
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/sample.py:118]: I cleaned it up; and replaced pick bond by grow chain

File: cgsmiles/sample.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        if self.all_atom:
+            rebuild_h_atoms(self.molecule)
+
+        # sort the atoms
+        self.molecule = sort_nodes_by_attr(self.molecule, sort_attr=("fragid"))
+
+        # in all-atom MD there are common naming conventions
+        # that might be expected and hence we set them here
+#        if self.all_atom:
+#            set_atom_names_atomistic(self.molecule)
```
[the remark this answers is not in the document]
Comment (fgrunewald) [reply cgsmiles/sample.py:134]: set_atom_names_atomistic requires a meta-graph that we don't build; I need to fix that

File: cgsmiles/sample.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pckroon) [reply cgsmiles/sample.py]: That sounds great

File: cgsmiles/sample.py [hunk 7]
```diff
@@ -7,133 +7,240 @@
...
-            of number of monomers
-        fragment_probabilities: dict[str, float]
-            dictionary connecting probability
-            to fragment names
+        fragment_dict: dict[str, nx.Graph]
+            a dict of fragment graphs at one resolution. Each graph must have
+            the same attributes as returned by the `cgsmiles.read_fragments`
+            function.
+        target_weight: int
+            the target molecular weight in unit of number of monomers
```
Comment (pckroon) [inline cgsmiles/sample.py:35]:
```suggestion
            the target molecular weight
```
?

File: cgsmiles/sample.py [hunk 8]
```diff
@@ -7,133 +7,240 @@
...
+        #    form the new bond
+        probs = np.array([self.bonding_probabilities[bond_type[:-1]] for bond_type in self.current_open_bonds])
+        probs = probs / sum(probs)
+        # 2. pick a random bonding descriptor according to these probs
         bonding = np.random.choice(list(self.current_open_bonds.keys()), p=probs)
-        source_node = np.random.choice(self.current_open_bonds[bonding])
-        compl_bonding = _find_complementary_bonding_descriptor(bonding)
+        # 3. get a corresponding node; it may be that one descriptor is found on
+        #    several nodes
+        random.seed(a=seed)
```
Comment (pckroon) [inline cgsmiles/sample.py:115]: This needs to move to the `__init__`, because the descriptor is picked *before* this

File: cgsmiles/sample.py [hunk 9]
```diff
@@ -7,133 +7,240 @@
...
+        source_node = random.choice(self.current_open_bonds[bonding])
+        # 4. get the complementary matching bonding descriptor
+        compl_bonding = find_complementary_bonding_descriptor(bonding)
+        # 5. pick a new fragment that has such bonding descriptor
+        random.seed(a=seed)
```
Comment (pckroon) [inline cgsmiles/sample.py:120]:
```suggestion
```
Seed only once

File: cgsmiles/sample.py [hunk 10]
```diff
@@ -7,133 +7,240 @@
...
+            the fragment by name to consider
+        fragid: int
+            the id of the fragment
+
+        Returns
+        -------
+        nx.Graph
+        """
+        term_prob = self.termination_probabilities.get(fragname, 0)
+        random.seed(a=seed)
```
Comment (pckroon) [inline cgsmiles/sample.py:151]:
```suggestion
```

File: cgsmiles/sample.py [hunk 11]
```diff
@@ -7,133 +7,240 @@
...
+        str
+            the fragment name of the added fragment
         """
-        probs = [self.bonding_probabilities[bond_type] for bond_type in self.current_open_bonds]
+        # 1. get the probabilties of any bonding descriptor on the chain to
+        #    form the new bond
+        probs = np.array([self.bonding_probabilities[bond_type[:-1]] for bond_type in self.current_open_bonds])
+        probs = probs / sum(probs)
+        # 2. pick a random bonding descriptor according to these probs
         bonding = np.random.choice(list(self.current_open_bonds.keys()), p=probs)
```
Comment (pckroon) [inline cgsmiles/sample.py:112]: Note that this uses the np random, which you don't seed

File: cgsmiles/sample.py [hunk 12]
```diff
@@ -7,133 +7,240 @@
...
+            the id of the fragment
+
+        Returns
+        -------
+        nx.Graph
+        """
+        term_prob = self.termination_probabilities.get(fragname, 0)
+        random.seed(a=seed)
+        # probability check for termination
+        if random.random() < term_prob:
```
Comment (pckroon) [inline cgsmiles/sample.py:153]:
```suggestion
        if random.random() <= term_prob:
```

File: cgsmiles/sample.py [hunk 13]
```diff
@@ -7,133 +7,240 @@
...
+            # check if there are more open bonding descriptors
+            # if the number is the same as would get removed
+            # then we are not on a branch
+            active_bonds = nx.get_node_attributes(molecule, 'bonding')
+            target_nodes = [ node for node in active_bonds if molecule.nodes[node]['fragid'] == fragid]
```
Comment (pckroon) [inline cgsmiles/sample.py:158]:
```suggestion
            target_nodes = [node for node in active_bonds if molecule.nodes[node]['fragid'] == fragid]
```

File: cgsmiles/sample.py [hunk 14]
```diff
@@ -7,133 +7,240 @@
...
+            if len(target_nodes) < len(active_bonds):
+                for node in target_nodes:
+                    del molecule.nodes[node]['bonding']
```
Comment (pckroon) [inline cgsmiles/sample.py:161]: This does make it impossible to figure out termini at a later point in life.
How about creating a set of "terminated" nodes that still have a bonding attribute, but can no longer be used by this sampler? I.e, `self.current_open_bonds -= self.terminated_nodes`

File: cgsmiles/sample.py [hunk 15]
```diff
@@ -7,133 +7,240 @@
...
         """
-        Sample one molecule given a fragment string and a target molecular
-        weigth.
+        From a list of cgsmiles fragment graphs generate a new random molecule
+        according by stitching them together.
+
+        Parameters
+        ----------
+        target_weight
+            the weight of the polymer to generate
```
Comment (pckroon) [inline cgsmiles/sample.py:173]: Maximum or minimum?
Example, target_weight is 10, current weight is 9. Do I add a weight 2 fragment or not?

File: cgsmiles/sample.py [hunk 16]
```diff
@@ -7,133 +7,240 @@
...
+        -------
+        nx.Graph
+            the graph of the molecule
         """
+        molecule = nx.Graph()
         if self.start:
             fragment = self.fragment_dict[self.start]
         else:
             # intialize the molecule; all fragements have the same probability
+            random.seed(a=seed)
```
Comment (pckroon) [inline cgsmiles/sample.py:185]:
```suggestion
```

**Review by pckroon — CHANGES_REQUESTED** [review]
I like most of the changes, but I'm not happy with how you seed the random generator. You should seed it only once, at the start of your process. Otherwise your seeded random will not be random enough.
You can also, for reproducibility, generate a seed if None is given (`time.time_ns()`) and log that somewhere the user can find it.

File: cgsmiles/sample.py (hunk 9 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:120]: that does not work; it only works if seed is called directly before the execution of the random statement

Comment (pckroon) [reply cgsmiles/sample.py:120]: Depends on what you call "works".
Generating a random number affects/changes the random state of the generator. This is intentional.
What you want is that after seeding (once), it generates a deterministic sequence of pseudo-random numbers.

Comment (pckroon) [reply cgsmiles/sample.py:120]: Or even better, make a `self._rng = random.Random(seed)` in the `__init__`, and then call `self._rng.choice` etc.

File: cgsmiles/sample.py (hunk 8 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:115]: done

File: cgsmiles/sample.py (hunk 14 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:161]: yes it does; I've adjusted the termination code such that possible end-groups are added before the bondings are removed.

File: cgsmiles/cgsmiles_utils.py [hunk 17]
```diff
@@ -16,14 +16,18 @@ def find_complementary_bonding_descriptor(bonding_descriptor):
         compl = bonding_descriptor
     return compl
 
-def find_open_bonds(molecule):
+def find_open_bonds(molecule, target_nodes=None):
```
Comment (pckroon) [inline cgsmiles/cgsmiles_utils.py:19]: Docstring does not describe arguments

File: cgsmiles/cgsmiles_utils.py [hunk 18]
```diff
@@ -16,14 +16,18 @@ def find_complementary_bonding_descriptor(bonding_descriptor):
...
     """
     Collect all nodes which have an open bonding descriptor and store
     them as keys with a list of nodes as values.
     """
+    if target_nodes is None:
+        target_nodes = list(molecule.nodes)
```
Comment (pckroon) [inline cgsmiles/cgsmiles_utils.py:25]:
```suggestion
        target_nodes = molecule
```
You only use it to check presence of items anyway

File: cgsmiles/pysmiles_utils.py [hunk 19]
```diff
@@ -1,7 +1,7 @@
 import networkx as nx
 import pysmiles
 
-def compute_mass(molecule):
+def compute_mass(input_molecule):
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py:4]: Update docstring

File: cgsmiles/pysmiles_utils.py [hunk 20]
```diff
@@ -15,6 +15,11 @@ def compute_mass(molecule):
     float
         the atomic mass
     """
+    molecule = input_molecule.copy()
+    print(molecule.nodes(data=True))
```
Comment (pckroon) [inline cgsmiles/pysmiles_utils.py:19]:
```suggestion
```

File: cgsmiles/sample.py [hunk 21]
```diff
@@ -17,12 +17,13 @@ class MoleculeSampler:
...
                  bonding_probabilities,
+                 branch_term_probs=None,
+                 terminal_fragments=[],
+                 bond_term_probs=None,
                  fragment_masses=None,
-                 termination_probabilities=None,
-                 start=None,
-                 all_atom=True):
+                 all_atom=True,
+                 seed=None):
```
Comment (pckroon) [inline cgsmiles/sample.py:26]: Update docstring

File: cgsmiles/sample.py [hunk 22]
```diff
@@ -39,31 +40,31 @@ def __init__(self,
...
+            is present at the terminal
         start: str
             fragment name of the fragment to start with
         all_atom: bool
             if the fragments are all-atom resolution
         seed: int
             set random seed for all processes; default is None
         """
+        # first initalize the random number generator
+        random.seed(a=seed)
```
Comment (pckroon) [inline cgsmiles/sample.py:57]:
```suggestion
        # first initalize the random number generator
        if seed is None:
            seed = time.time_ns()
            logger.info("Your random seed is %i", seed)
        random.seed(a=seed)
```

File: cgsmiles/sample.py [hunk 23]
```diff
@@ -106,30 +117,58 @@ def grow_chain(self, molecule, seed=None):
...
+        descriptors of all nodes will be removed.
+
+        Parameters
+        ----------
+        molecule: nx.Graph
+            the molecule graph
+        fragid: int
+            the id of the fragment
+        """
+        target_nodes = [node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid']]
```
Comment (pckroon) [inline cgsmiles/sample.py:155]:
```suggestion
        target_nodes = {node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid']}
```
A set should be marginally faster

**Review by pckroon — CHANGES_REQUESTED** [review]

File: cgsmiles/sample.py (hunk 15 above)
Comment (fgrunewald) [reply cgsmiles/sample.py]: minimum; docstring is updated

File: cgsmiles/sample.py (hunk 22 above)
Comment (pckroon) [reply cgsmiles/sample.py]: Those can always adjust the loglevel from specifically this module.

File: cgsmiles/sample.py [hunk 24]
```diff
@@ -10,17 +11,134 @@
...
+
+    One can also make a blocky copolymer but steer the probability of having
+    head-to-head tail-to-tail vs head-to-tail addition. This is done by labeling
+    the bonding descriptors A-D and providing the probability that a certain
+    bonding descriptor is used to extend the polymer.
+
+    >>> cgsmiles_str = "{#PMMA=[>A]C(C)C[<B]C(=O)OC,#PS=[>C]CC[<D]c1cccc1}"
+    >>> polymer_reactivites = {">A":0.4, "<B":0.1, ">C": 0.4, "<D": 0.1}
+    >>> sampler = MoleculeSampler.from_string(cgsmiles_str)
+    >>> polymer_graph = sampler.sample(target_weight=100000)
```
Comment (pckroon) [inline cgsmiles/sample.py:77]: I'm not sure I understand this example. Once I have a PMMA with a free `<B`, isn't it just as likely to take `>A` (=PMMA) as it is `>C` (=PS)?
Besides a block copolymer, could you add an example to create an alternating copolymer (using the chances)

Also, whitespace after the `:` in `">A":0.4, "<B":0.1`

File: cgsmiles/sample.py [hunk 25]
```diff
@@ -10,17 +11,134 @@
...
+
+    One can also make a blocky copolymer by providing the conditional
+    probabilities. To distinguish the descriptors on PMMA vs PS we can
+    label them with a alphanumeric string.
+
+    >>> cgsmiles_str = "{#PMMA=[$A]C(C)C[$B]C(=O)OC,#PS=[$C]CC[$D]c1cccc1}"
+    >>> polymer_reactivites = {">A":0.5, "<A":0, ">B": 0.5, "<B": 0.0}
```
Comment (pckroon) [inline cgsmiles/sample.py:84]: But, these bonding descriptors are no longer in the polymer?

File: cgsmiles/sample.py [hunk 26]
```diff
@@ -10,17 +11,134 @@
...
+    in combination with the above mentioned probabilities can be provided.
+
+    >>> cgsmiles_str="{#PMA=[>]CC[<]C(=O)OC[>A],#PEG=[<A]COC[>A],#OH=[<A]O}"
+    >>> sampler = MoleculeSampler.from_fragment_string(cgsmiles_str,
+                                                       branch_term_probs={"PEG": 0.3, "PMA":0.0},
+                                                       terminal_fragments=['OH'],
+                                                       bonding_probabilities={'<': 0.01, '>': 0.01, '>A': 0.99, '<A':0.99},
+                                                       bond_term_probs={'<A': 1., '>A':1., '>': 0, '<': 0},
+                                                       all_atom=True)
+    >>> molecule = sampler.sample(target_weight=5008)
```
Comment (pckroon) [inline cgsmiles/sample.py:113]: Not sure I like the whole terminal thing. Wouldn't the following do (roughly) the same?
```
cgsmiles_str="{#PMA=[>]CC[<]C(=O)OC[>A],#PEG=[<A]COC[>A],#OH=[<T]O}"
sampler = MoleculeSampler.from_fragment_string(cgsmiles_str,
                                                       polymer_reactivities={'<': 0.01, '>': 0.01, '>A': 0.99, '<A':0.99},
                                                       fragment_reactivities={
'>': {'<': 1}, '<': {'>': 1}, '>A': {'<A': 0.7, '<T': 0.3}}  # etc.
                                                       all_atom=True)
```
The whole separation/relation between bonding probabilities and polymer reactivities is not super clear to me. What does the first allow you to do you cannot do with the second?

File: cgsmiles/sample.py [hunk 27]
```diff
@@ -10,17 +11,134 @@
...
+
+    This combination results into a dense brush where very PMA molecule has a PEG
+    branch whose length is determined by a 1 in 3 probability to terminate.
+
+    In contrast one can also generate spares branched molecules. For instance,
```
Comment (pckroon) [inline cgsmiles/sample.py:118]:
```suggestion
    In contrast one can also generate sparse branched molecules. For instance,
```

File: cgsmiles/sample.py [hunk 28]
```diff
@@ -10,17 +11,134 @@
...
+    Dextran at low molecular weights is mainly a linear alpha 1-6 polyglucose
+    polymer. Occasionally, a 1-4 branch extends from it which has lengths of 2-4.
+    At the Martini coarse-grained level Dextran is described by three particles
+    A, B, and C. The alpha 1-6 bonds are between A and C whereas the 1-4 bonds
+    are between A and B.
+
+    >>> cgsmiles_str="{#GLC=[$A][#A]1[#B][$B][#C]1[$C]}"
+    >>> sampler = MoleculeSampler.from_fragment_string(cgsmiles_str,
+                                                       fragment_masses={'GLC': 165},
+                                                       bonding_probabilities={'$A': 0.8, '$C': 0.2, '$B': 0.2},
```
Comment (pckroon) [inline cgsmiles/sample.py:128]: This does not sum to 1?

File: cgsmiles/sample.py [hunk 29]
```diff
@@ -152,15 +290,17 @@ def terminate_fragment(self, molecule, fragid):
...
+        # if terminal fragment bonding probabilities are given; add them here
+        if self.terminal_reactivities:
             self.add_fragment(molecule,
                               open_bonds,
                               self.terminals_by_bonding,
-                              self.bond_term_probs)
+                              self.terminal_reactivities,
+                              self.fragment_reactivities)
             fragid += 1
+            target_nodes.update({node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid']})
```
Comment (pckroon) [inline cgsmiles/sample.py:303]: nitpick
```suggestion
            target_nodes.update(node for node in molecule.nodes if fragid in molecule.nodes[node]['fragid'])
```

**Review by pckroon — CHANGES_REQUESTED** [review]
For the Sampler examples, I'm not super clear on what the difference between \*reactivities and \*probabilities is. Also, for some examples you do `from_string`, for other `from_fragment_string`. Is this intentional?
Lastly, does it make sense to completely replace the terminal logic with the fragment reactivities?

File: cgsmiles/resolve.py [hunk 30]
```diff
@@ -23,14 +23,23 @@ def compatible(left, right):
...
-def match_bonding_descriptors(source, target, bond_attribute="bonding"):
+    if legacy:
+        if left == right and left[0] not in '> <':
+            return True
+        l, r = left[0], right[0]
+        if (l, r) == ('<', '>') or (l, r) == ('>', '<'):
+            return left[1:] == right[1:]
+        return False
+    else:
+        if left[0] == right[0] == '$' or left[0] == right[0] == '!':
```
[the remark this answers is not in the document]
Comment (pckroon) [reply cgsmiles/resolve.py:34]: Cool.

File: cgsmiles/sample.py (hunk 26 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:113]: CG representation where you don't have a terminal fragment.

File: cgsmiles/sample.py (hunk 25 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:84]: uups; my bad I did not update the reactivities

File: cgsmiles/sample.py (hunk 28 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:128]: does not have to but I guess might be better for the example purpose.

File: cgsmiles/sample.py (hunk 24 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:77]: Yep the example was somehow budge. I corrected it and moved it to the bottom of the basic examples

**Comment by fgrunewald** [comment]
@pckroon All reactivities are probabilities. Reactivities is used when the decision is based on a bonding descriptor. I've added a clarification in the docs. 

they should all be `from_fragment_string` but some examples were older from before 

Not entirely because at the CG level branch termination cannot be determined by the bonding descriptor when no end-group is added. Otherwise, yes one could do as suggest. Of course one other option is to simply define an empty or fake terminal. But not sure that makes it easier.

File: cgsmiles/resolve.py [hunk 31]
```diff
@@ -414,6 +434,11 @@ def from_graph(cls, cgsmiles_str, meta_graph, last_all_atom=True):
             fragname attribute set.
         last_all_atom: bool
             if the last resolution is all-atom and is read using pysmiles
+        legacy: bool
+            which syntax convention to use for matching the bonding descriptors.
+            Legacy syntax adheres to the BigSmiles convention. Default syntax
+            adheres to CGSmiles convention. A more detailed explanation can be
+            found in the __init__ method.
```
Comment (pckroon) [inline cgsmiles/resolve.py:441]: You can turn this into an automatic sphinx link (somehow). I don't have time to figure out how though.

File: cgsmiles/resolve.py [hunk 32]
```diff
@@ -449,6 +475,11 @@ def from_fragment_dicts(cls, cgsmiles_str, fragment_dicts, last_all_atom=True):
             function.
         last_all_atom: bool
             if the last resolution is all-atom and is read using pysmiles
+        legacy: bool
+            which syntax convention to use for matching the bonding descriptors.
+            Legacy syntax adheres to the BigSmiles convention. Default syntax
+            adheres to CGSmiles convention. A more detailed explanation can be
+            found in the __init__ method.
```
Comment (pckroon) [inline cgsmiles/resolve.py:482]: Ditto

File: cgsmiles/sample.py [hunk 33]
```diff
@@ -43,11 +46,16 @@ class MoleculeSampler:
...
 
     In addition to just randomly creating a molecule reactivities and termination
-    probabilities can be provided to steer the outcome. The sampler features a
-    two level connectivity determination. First using the `polymer_reactivities`
-    an open bonding descriptor from the growing polymer molecule is selected
-    according to the probabilities provided. Subsequently, a fragment is randomly
-    chosen that has a matching complementary bonding descriptor.
+    probabilities can be provided to steer the outcome. We use the term and variable
+    name `reactivity` whenever referring to a probability that a certain bonding
+    descriptor is used to highlight the fact that the probabilistic behviour is driven
```
Comment (pckroon) [inline cgsmiles/sample.py:51]:
```suggestion
    descriptor is used to highlight the fact that the probabilistic behaviour is driven
```

**Review by pckroon — CHANGES_REQUESTED** [review]
Thanks for the clarification. The docstring(s) also help me :)
Still found a typo though.

As for the terminals, I don't have a super strong preference for one or the other. I just know from experience that the more complex an API is, the harder it is to use. It requires more mental space to keep track of all the arguments, their different formats/types/conventions, and how it all interacts.

File: cgsmiles/sample.py [hunk 34]
```diff
@@ -16,21 +16,29 @@
...
-def _set_bond_order_defaults(bonding_dict):
-    default_dict = {}
-    for bond_operator, prob in bonding_dict.items():
-        # we need to patch the bond order space
-        if not bond_operator[-1].isdigit():
-            bond_operator += '1'
-        default_dict[bond_operator] = prob
-    return default_dict
+def _set_bond_order_defaults(bonding):
+    if type(bonding) == dict:
```
Comment (pckroon) [inline cgsmiles/sample.py:27]:
```suggestion
    if isinstance(bonding, dict):
```

File: cgsmiles/sample.py [hunk 35]
```diff
@@ -16,21 +16,29 @@
...
+                bond_operator += '1'
+            default_dict[bond_operator] = prob
+        return default_dict
+    else:
+        default_list = []
+        for bond_operator in bonding:
+            if not bond_operator[-1].isdigit():
+                bond_operator += '1'
+            default_list.append(bond_operator)
+        return default_list
```
Comment (pckroon) [inline cgsmiles/sample.py:41]: Is there a way to normalize this to either a list or a dict?

File: cgsmiles/sample.py [hunk 36]
```diff
@@ -202,9 +212,10 @@ def __init__(self,
             if not key[-1].isdigit():
                 key += '1'
             self.fragment_reactivities[key] = _set_bond_order_defaults(probs)
-        self.terminal_reactivities = _set_bond_order_defaults(terminal_reactivities)
-        self.branch_term_probs = branch_term_probs
+
         self.all_atom = all_atom
+        # make sure to get bond order defaults
+        self.terminal_bonds = _set_bond_order_defaults(terminal_bonds)
```
Comment (pckroon) [inline cgsmiles/sample.py:218]: Normalize `terminal_bonds` to be a dict here.

**Review by pckroon — CHANGES_REQUESTED** [review]
Still found some small points :)

File: cgsmiles/sample.py (hunk 36 above)
Comment (fgrunewald) [reply cgsmiles/sample.py:218]: but that makes no sense what would be the corresponding key:value?

Comment (pckroon) [reply cgsmiles/sample.py:218]: My brain is mushy, what does the dict look like if user provided?

Comment (fgrunewald) [reply cgsmiles/sample.py:218]: terminal_bonds is a list with terminal bonding descriptors not a dict

Comment (pckroon) [reply cgsmiles/sample.py:218]: ... but `_set_bond_order_defaults` can also deal with dicts

Comment (fgrunewald) [reply cgsmiles/sample.py:218]: yes the dict are the fragment_reactivities:

{'bond_1': {bond1: 0.0, bond2: 1.0},
 'bond_2': {bond2: 0.0 bond1: 1.0}}

Comment (fgrunewald) [reply cgsmiles/sample.py:218]: set bond_order just adds a 1 to all bonding descriptors if they don't end in a digit and I needed it to work for a list and double dict. Hence that private function

**Review by pckroon — APPROVED** [review]

**Pull request merged**